src/realm/sync/subscriptions.cpp

```diff
@@ -798,6 +798,47 @@ std::vector<SubscriptionSet> SubscriptionStore::get_pending_subscriptions() cons
     return subscriptions_to_recover;
 }
 
+void SubscriptionStore::notify_all_state_change_notifications(Status error)
+{
+    std::unique_lock<std::mutex> lk(m_pending_notifications_mutex);
+    m_pending_notifications_cv.wait(lk, [&] {
+        return m_outstanding_requests == 0;
+    });
+
+    auto to_finish = std::move(m_pending_notifications);
+    lk.unlock();
+
+    // Just complete/cancel the pending notifications - this function does not alter the
+    // state of any pending subscriptions
+    for (auto& req : to_finish) {
+        req.promise.set_error(error);
+    }
+}
+
+void SubscriptionStore::terminate()
+{
+    auto tr = m_db->start_write();
+    auto sub_sets = tr->get_table(m_sub_set_table);
+    sub_sets->clear();
+    auto zero_sub = sub_sets->create_object_with_primary_key(Mixed{int64_t(0)});
+    zero_sub.set(m_sub_set_state, static_cast<int64_t>(SubscriptionSet::State::Pending));
+    zero_sub.set(m_sub_set_snapshot_version, tr->get_version());
+    tr->commit();
+
+    std::unique_lock<std::mutex> lk(m_pending_notifications_mutex);
```

Comment:
nit: there is some code duplication with the method above. Could consider refactoring it